packages/git/src/common/git-watcher.ts

```diff
@@ -0,0 +1,188 @@
+/*
+* Copyright (C) 2017 TypeFox and others.
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+import { injectable, inject } from "inversify";
+import { JsonRpcServer, JsonRpcProxy } from '@theia/core';
+import { Repository, WorkingDirectoryStatus } from './model';
+import { Disposable, DisposableCollection, Emitter } from '@theia/core/lib/common';
+
+/**
+ * An event representing a Git change.
+ */
+export interface GitEvent {
+
+    /**
+     * The source `Git` repository where the event belongs to.
+     */
+    readonly source: Repository;
+
+    /**
+     * The old `Git` source. Might be missing.
+     */
+    readonly oldSource?: Repository;
```

Comment:
Sure, we have to listen to two kinds of events: 
 - A new repository appears or gets removed from the workspace.
 - The other type of the event is when the status has changed in a repository.

Both of the event types is a subtype of the `GitEvent` type. Since a repository "deletion" has to be distinguished from a repository "creation" event, we have the source, which stands for the creation event, and we have the "oldSource" which is for the deletion.

Of course, we can have two different types and have the union type of them, I did not want to do that on purpose but we can change.